### PR TITLE
Nova 4 Compatibility

### PR DESCRIPTION
**Pull request by marcfilipan**
Breaking changes
  - Removed selectFirstTab, first tab is always displayed first.
  - Even if you have other panels, tabs will always show up first and has the toolbar.
  - TabsOnEdit is gone and non relational tabs will simply always display on edit.
  - I don't use dusk, so didn't check the tests for it either, they might be broken.
  - To make HasMany, BelongsToMany, HasManyThrough, HasOneThrough and MorphToMany work, you'd need to use their respective fields within this package to remove nova 4's AsPanel method from the fields.
  - Added Eminiarts\Tabs\Traits\HasTabs to overwrite Nova 4's panelsWithDefaultLabel method in Laravel\Nova\ResolveFields to enable tabs on edit pages.
  - Moved Eminiarts\Tabs\ActionsInTabs to Eminiaarts\Tabs\Traits\HasActionsInTabs 
  - Removed before and after icons
  - Removed html title
  - Removed custom tab classes

It's quite basic, but it works.

### CODE REVIEW

**Comment by marcfilipan** [comment]
Forgot to mention that to keep it responsive, it switches to a select dropdown when screen size is smaller than 768px

**Comment by marcfilipan** [comment]
Crynonone pointed me towards a nicer solution for the relationship tabs, I'll dive into it tomorrow.

**Comment by ibrunotome** [comment]
> Crynonone pointed me towards a nicer solution for the relationship tabs, I'll dive into it tomorrow.

Hi @marcfilipan, could I ask you to test the tabs with "optimistdigital/nova-settings": "^4.0.2" ?

Works perfectly on nova v3 but is not working with your fork. Nothing is displayed.

**Comment by ackermann** [comment]
@marcfilipan I've just rebased your changes to [my branch](https://github.com/ackermann/nova-tabs) – would love for you to check out [PR #1](https://github.com/marcfilipan/nova-tabs/pull/1) and give me a feedback on the style direction.

**Comment by marcfilipan** [comment]
> > Crynonone pointed me towards a nicer solution for the relationship tabs, I'll dive into it tomorrow.
> 
> Hi @marcfilipan, could I ask you to test the tabs with "optimistdigital/nova-settings": "^4.0.2" ?
> 
> Works perfectly on nova v3 but is not working with your fork. Nothing is displayed.

From what I see the tabs parts are commented out in the 4.* versions of nova-settings.

File: prettier.config.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      files: 'resources/css/**/*.css',
+      options: {
+        tabWidth: 2,
+      },
+    },
+    {
+      files: [
+        'resources/js/fields.js',
+        'resources/js/components.js',
+        'resources/js/router/routes.js',
```
Comment (crynobone) [inline prettier.config.js:25]:
```suggestion
```

Comment (crynobone) [reply prettier.config.js:25]: Vue Router should no longer work on Nova 4

File: resources/js/components/DetailTabs.vue [hunk 2]
```diff
@@ -1,211 +1,261 @@
...
+    },
+
+    /**
+     * Sort the tabs object by their respective positions using lodash
+     *
+     * @param tabs
+     * @returns {object}
+     */
+    getSortedTabs(tabs) {
+      return _.orderBy(tabs, [c => c.position], ['asc']);
```
Comment (crynobone) [inline resources/js/components/DetailTabs.vue:253]:
```suggestion
      return orderBy(tabs, [c => c.position], ['asc']);
```

File: resources/js/components/DetailTabs.vue [hunk 3]
```diff
@@ -1,211 +1,261 @@
...
 </template>
 
 <script>
-import BehavesAsPanel from 'laravel-nova/src/mixins/BehavesAsPanel';
-import TabTitle from './TabTitle';
-import { changeActiveTab } from '@/util/tab-updater';
+import BehavesAsPanel from '../../../vendor/laravel/nova/resources/js/mixins/BehavesAsPanel';
+import Heading from '../../../vendor/laravel/nova/resources/js/components/Heading.vue';
+import Card from '../../../vendor/laravel/nova/resources/js/components/Card.vue';
+import _ from "lodash";
```
Comment (crynobone) [inline resources/js/components/DetailTabs.vue:80]:
```suggestion
import orderBy from 'lodash/orderBy';
```

File: resources/js/components/FormTabs.vue [hunk 4]
```diff
@@ -1,233 +1,290 @@
...
-    HandlesValidationErrors,
-    InteractsWithResourceInformation,
-} from 'laravel-nova';
-
-import {changeActiveTab} from '@/util/tab-updater';
-import TabTitle from './TabTitle';
+import BehavesAsPanel from '../../../vendor/laravel/nova/resources/js/mixins/BehavesAsPanel';
+import Heading from '../../../vendor/laravel/nova/resources/js/components/Heading.vue';
+import Card from '../../../vendor/laravel/nova/resources/js/components/Card.vue';
+import _ from "lodash";
```
Comment (crynobone) [inline resources/js/components/FormTabs.vue:93]:
```suggestion
import orderBy from 'lodash/orderBy';
```

File: resources/js/components/FormTabs.vue [hunk 5]
```diff
@@ -1,233 +1,290 @@
...
-    FormField,
-    HandlesValidationErrors,
-    InteractsWithResourceInformation,
-} from 'laravel-nova';
-
-import {changeActiveTab} from '@/util/tab-updater';
-import TabTitle from './TabTitle';
+import BehavesAsPanel from '../../../vendor/laravel/nova/resources/js/mixins/BehavesAsPanel';
+import Heading from '../../../vendor/laravel/nova/resources/js/components/Heading.vue';
+import Card from '../../../vendor/laravel/nova/resources/js/components/Card.vue';
```
Comment (crynobone) [inline resources/js/components/FormTabs.vue:92]: Both `Heading` and `Card` should be available globally

File: resources/js/components/DetailTabs.vue [hunk 6]
```diff
@@ -1,211 +1,261 @@
...
+  </div>
 </template>
 
 <script>
-import BehavesAsPanel from 'laravel-nova/src/mixins/BehavesAsPanel';
-import TabTitle from './TabTitle';
-import { changeActiveTab } from '@/util/tab-updater';
+import BehavesAsPanel from '../../../vendor/laravel/nova/resources/js/mixins/BehavesAsPanel';
+import Heading from '../../../vendor/laravel/nova/resources/js/components/Heading.vue';
+import Card from '../../../vendor/laravel/nova/resources/js/components/Card.vue';
```
Comment (crynobone) [inline resources/js/components/DetailTabs.vue:79]: Both `Heading` and `Card` should be available globally

File: resources/js/components/FormTabs.vue [hunk 7]
```diff
@@ -1,233 +1,290 @@
...
+    },
+
+    /**
+     * Sort the tabs object by their respective positions using lodash
+     *
+     * @param tabs
+     * @returns {object}
+     */
+    getSortedTabs(tabs) {
+      return _.orderBy(tabs, [c => c.position], ['asc']);
```
Comment (crynobone) [inline resources/js/components/FormTabs.vue:282]:
```suggestion
      return orderBy(tabs, [c => c.position], ['asc']);
```

**Comment by marcfilipan** [comment]
> @marcfilipan may I ask why you removed the tailwind color config? With my setup it was necessary to make sure that the background had the correct grey tone.
> 
> On another note: I think you have replaced two „nt“‘s too much (fo_tabs_- instead of font-). nt was standing for NovaTabs, by the way.

Fixed the font issue, thanks. 

About the TW config, which shade of grey goes wrong for you and where? Perhaps I overlooked something.

**Comment by ackermann** [comment]
> About the TW config, which shade of grey goes wrong for you and where? Perhaps I overlooked something.

It‘s really (!) subtle, but without that config the card (tab panel) background in dark mode has the wrong (i.e. not the original Nova) grey tone, which is not default grey but blue grey (aka slate).

**Comment by marcfilipan** [comment]
> > About the TW config, which shade of grey goes wrong for you and where? Perhaps I overlooked something.
> 
> It‘s really (!) subtle, but without that config the card (tab panel) background in dark mode has the wrong (i.e. not the original Nova) grey tone, which is not default grey but blue grey (aka slate).

Is this better?

**Comment by ackermann** [comment]
> Is this better?

I can‘t test it right now but the config looks perfect – will check later.

**Comment by simchaba** [comment]
Nova 4 is working with Laravel 8 and PHP 7.4, your composer state php 8. Is there any real limitation on this or your composer.json could state "php": "^7.4|^8.0" ?

**Comment by ackermann** [comment]
@marcfilipan the colors look alright now, thanks for reverting the change!

I've just discovered a regression where a tab group that contains only relationship tabs is not rendered. Adding at least one non-relationship makes the tabs appear again, but in the wrong order.

**Comment by marcfilipan** [comment]
> Nova 4 is working with Laravel 8 and PHP 7.4, your composer state php 8. Is there any real limitation on this or your composer.json could state "php": "^7.4|^8.0" ?

Simply because I don't use it and don't care to test with it, I can enable it, but 



> @marcfilipan the colors look alright now, thanks for reverting the change!
> 
> I've just discovered a regression where a tab group that contains only relationship tabs is not rendered. Adding at least one non-relationship makes the tabs appear again, but in the wrong order.

I'll look into it, also discovered a bug with hasOne/morphOne creations, got that fixed now.

**Comment by marcfilipan** [comment]
@simchaba added 7.4, please try it

@ackermann fixed the issue and merged your PR.

**Comment by marcfilipan** [comment]
@bajramemini or @RVxLab 

Are any of you guys around still?

**Comment by simchaba** [comment]
@marcfilipan just tested with PHP 7.4 working just fine for me.

**Comment by RVxLab** [comment]
I’m not working on the Nova 4 version as I don’t have a license. See #180

**Comment by ackermann** [comment]
@bajramemini @RVxLab do you think that (definitely) forking this repo would be the better approach to building and maintaining the N4 version than pull requests to this repo? 
Another option would be promoting @marcfilipan (or other volunteers) to maintainer – or transferring the project completely.

**Comment by RVxLab** [comment]
Transferring the project is out of my control as I'm collaborator. As for forking it that'd mostly going to depend on whether @bajramemini is going to be active to handle anything.

However, with the owner being inactive in this repo going with the fork route may be better.

I'll see if I can reach him directly over email.

**Comment by marcfilipan** [comment]
> @marcfilipan just tested with PHP 7.4 working just fine for me.

Awesome, thanks for confirming

**Comment by bajramemini** [comment]
Hello everyone.

Hello @marcfilipan 

Thank you very much for your PR. 

We've been busy with other projects and non Nova-related stuff.

We would be happy to add some Collaborators/Maintainers to this package.

@marcfilipan Would you like to help to maintain this package?

If so, I would be happy to add you as a collaborator. 

I didn't have the chance to review the changes yet and test them out. I think, I can tackle this next week and merge your PR if everything is fine. 

I'll also ask Acato Devs if they want to participate in collaborating/maintaining this package. #183

**Comment by marcfilipan** [comment]
Hi @bajramemini,

I can help maintain the package, but preferably together with @ackermann and @crynobone, as they've helped on it quite a bit.

**Comment by maloun96** [comment]
Can we release it?

**Comment by ziming** [comment]
Anyway to have TabsOnEdit back?

**Comment by marcfilipan** [comment]
> Anyway to have TabsOnEdit back?

They work on edit at the moment by default, currently working on making it optional as well as an optional wizard/step like style.

**Comment by bajramemini** [comment]
I just added you to the repository. Please create a new Release for Nova 4 (v2.0.0) when you are ready! 👍 thanks for your collaboration!

**Comment by sineld** [comment]
Merge please.

**Pull request merged**